peps/pep-0733.rst

```diff
@@ -0,0 +1,660 @@
+exposed in the C API as an actual pointer rather than a handle. The
+address of an object serves as its ID and is used for comparison,
+and this complicates matters for alternative Python implementations
+that move objects during GC
+[`Issue 37 <https://github.com/capi-workgroup/problems/issues/37>`__].
+
+A separate issue is that object references are opaque to the runtime,
+discoverable only through calls to ``tp_traverse``/``tp_clear``,
+which have their own purposes. If there was a way for the runtime to
+know the structure of the object graph, and keep up with changes in it,
+this would make it possible for alternative implementations to implement
+different memory management schemes
+[`Issue 33 <https://github.com/capi-workgroup/problems/issues/33>`__].
+
+Object Reference Management
+---------------------------
+
+There are C API functions that return borrowed references, and
+functions that steal references to arguments, but there isn't a
+naming convention that makes this obvious, so this is error prone
+[`Issue 8 <https://github.com/capi-workgroup/problems/issues/8>`__
+and `Issue 52 <https://github.com/capi-workgroup/problems/issues/52>`__].
+The terminology used to describe these situations in the documentation
+can also be improved
+[`Issue 11 <https://github.com/capi-workgroup/problems/issues/11>`__].
+
+A more radical change is necessary in the case of functions that
+return borrowed references (such as ``PyList_GetItem``)
+[`Issue 5 <https://github.com/capi-workgroup/problems/issues/5>`__ and
+`Issue 21 <https://github.com/capi-workgroup/problems/issues/21>`__]
+or pointers to parts of the internal structure of an object
+(such as ``PyBytes_AsString``)
+[`Issue 57 <https://github.com/capi-workgroup/problems/issues/57>`__].
+In both cases, the reference/pointer is valid for as long as the
+owning object holds the reference, but this time is hard to reason about.
+Such functions should not exist in the API without a mechanism that can
+make them safe.
+
+For containers, the API is currently missing bulk operations on the
+references of contained objects. This is particularly important for
+a stable ABI where ``INCREF`` and ``DECREF`` cannot be macros, making
+bulk operations expensive when implemented as a sequence of function
+calls
+[`Issue 15 <https://github.com/capi-workgroup/problems/issues/15>`__].
+
+Type Definition and Object Creation
+-----------------------------------
+
+The C API has functions that make it possible to create incomplete
+or inconsistent Python objects, such as ``PyTuple_New`` and
+``PyUnicode_New``. This causes problem when the object is tracked
+by GC or its ``tp_traverse``/``tp_clear`` functions are called.
+Such functions should be removed from the C API. Related functions,
+such as ``PyTuple_SetItem`` which is used to modify a partially
+initialized tuple, should also be removed (tuples are immutable
+once fully initialized)
+[`Issue 56 <https://github.com/capi-workgroup/problems/issues/56>`__].
```

Comment:
I expect this to be contentious for some users. Just because a tuple is immutable **at the Python level** doesn't mean all of its contents need exist up front elsewhere in a form that an acceptable non-partially-formed object API building a tuple could use. The problem statement here is real, but the solution statement of "just get rid of these APIs, effectively telling people who want efficient tuple creation they're SOL" doesn't smell right.

I'm not claiming that we'll do any differently. But that we should expect objections from real users if we do so.

I see there is more useful active discussion on this in the linked issue (great!).  so just take this comment as an FYI about how I expect some readers to react...